USB Host: integrate class API [4: aggregated]

josuah opened the pull request:
Dependencies:
- https://github.com/zephyrproject-rtos/zephyr/pull/94590
- https://github.com/zephyrproject-rtos/zephyr/pull/94591
- https://github.com/zephyrproject-rtos/zephyr/pull/99591
- https://github.com/zephyrproject-rtos/zephyr/pull/99578
- https://github.com/zephyrproject-rtos/zephyr/pull/99334

This PR is only here to aggregate multiple other PRs together as a way to combine all "preview" features together.

**It is not meant to be merged, and will stay a draft.**
**You might be interested in basing your PR on top of API2 or even better: directly on top of `main`**

If you wish to add more commit to this PR, I would recommend to open a PR against the `main` branch, and add a message in this thread so I can pick the commit. This way, your change can be merged directly.

Commit:
tests: usb: uvc: connect USB device and host via UVB

Add a test to run the USB Video Class host support by using
the existing Zephyr USB Video Class device support.
This allows running implementing the host side from the device side.
A draft implementation of UVC is added leveraging this test.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: uvc: move the header definition to include/

Move the UVC header with all the definitions from the UVC standard to
share it between USB host and device class implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: uvc: move the header definition to include/

Move the UVC header with all the definitions from the UVC standard to
share it between USB host and device class implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: uvc: placeholder implementation

Loop through each of the VideoStreaming and VideoControl descriptor
to parse them. This is meant as a stub for the purpose of testing the
class API.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: add host status to the host context

Add a "struct usbh_status" that contains a bitmask of flags to keep
track of the global state of the host context, like done for the
device_next implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: class: implement more of the class/data struct fields

Add a "struct usbh_class_api" for the host implementation, and move all
the function poitners to it. Add more fields to "struct usbh_class_data".

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: introduce wrappers to access the class function pointers

Add API wrappers around the function pointers in struct usbh_class_api,
while also documenting the USB host class internal API.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: introduce usbh_class with init/remove functions

Add functions to probe/remove all classes as part of a new usbh_class.c
and a matching usbh_class.h. These functions are called from the function
usbh_init_device_intl() in usbh_core.c to initialize every class upon
connection of a device. Every class driver provide filters to match the
interfaces of the device.

Co-authored-by: Aiden Hu <weiwei.hu@nxp.com>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
uvc: device: move helper functions to a common directory

Move UVC helper functions to a file shared between UVC host and device.
The arrays are not visible anymore from either USB host or device, but
instead accessed through a front-end funciton.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
tests: usb: host: test basic API functionality

Add tests making sure the USB Host class APIs introduced build
and run as expected.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
tests: usb: host: test basic API functionality

Add tests making sure the USB Host class APIs introduced build
and run as expected.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: add copyright notice to linker script

Add missing copyright notice for the linker script to help with
check_compliance.py.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: device_next: uvc: fix public API include

The public API file <zephyr/usb/class/usbd_uvc.h> lacked an include
to <zephyr/drivers/video.h> making it fail depending on the order of
the includes.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

santhosh-c-c committed:
usb: host: device: Read descriptor after setting device address

Modify the USB device int sequence to read the device
descriptor only after setting a valid device address.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

AidenHu committed:
subsys: usb: host: support multiple devices under hub

When hub is used, need to consider about multiple
devices are attached.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu commented:
@josuah
I have tested the current host core by this newest [API4](https://github.com/zephyrproject-rtos/zephyr/pull/99775).
Now the detach/attach for hub1->hub2->device ecm can work based on the newest [hub_class](https://github.com/zephyrproject-rtos/zephyr/pull/99735/commits/4cf9c8c6dd743d57d7c3571f881a453cc1d44048). This indicates the current host core is fine. 

I have updated the usb host connection/disconnection by [#100072](https://github.com/zephyrproject-rtos/zephyr/pull/100072).
This PR is created based on this API4 and I have tested by my local environment. Need your help to view it. If the change is fine, hope you can merge this commit [#100072](https://github.com/zephyrproject-rtos/zephyr/pull/100072) to  [API4](https://github.com/zephyrproject-rtos/zephyr/pull/99775).

Furthermore, I also need your help to merge the commit [hub_class](https://github.com/zephyrproject-rtos/zephyr/pull/99735/commits/4cf9c8c6dd743d57d7c3571f881a453cc1d44048) to your [API4](https://github.com/zephyrproject-rtos/zephyr/pull/99775), which this will make me easier to use newest code. Here sorry about that previously I preferred [hub_class](https://github.com/zephyrproject-rtos/zephyr/pull/99735/commits/4cf9c8c6dd743d57d7c3571f881a453cc1d44048) as a standalone PR.

Commit:
usb: uvc: common: correct the size for UVC_PU_CONTRAST_CONTROL

The USB control size field was wrong for UVC_PU_CONTRAST_CONTROL.
Correct it to the correct value from the standard.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: use SPDX-FileCopyrightText prefix

Switch to the SPDX-FileCopyrightText prefix for all sources in USB Host
class.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: uvc: rename "USBD" to "USB" in CONFIG_USBD_VIDEO_MAX_FRMIVAL

Allow the MAX_FRMIVAL Kconfig to be used for both device and host
implementation as the header is in subsys/usb/common/.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: add debug logs for the configuration descriptor

Every USB host class and platform will have to select the appropriate
size for the configurtion descriptor, and the defaults will not be enough
for some classes that encode a lot of information in the configuration
descriptor. Add logging to debug this.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: improve safety of configuration descriptor parsing

Make sure that there is at least enough room for the header before
dereferencing the fields such as bLength.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: add a nil entry at the end of configuration descriptor

Allocate a larger size to keep room for a nil descriptor at the end of
the configuration descriptor. This is used to walk through the list of
descriptors until this empty descriptor happens.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

jfischer-no committed:
usb: host: add helpers to parse USB descriptors

Add helpers to parse USB descriptors.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

jfischer-no committed:
usb: host: add functions to initialize and probe class drivers

Add basic functions to initialize and probe class drivers.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>
